### PR TITLE
feat: dApp staking via EVM wallets

### PR DESCRIPTION
**Pull request by impelcrypto**
**Pull Request Summary**

* feat: enabled dApp staking via EVM wallets
  * claiming rewards is enabled on Shibuya network only until runtime updating
  * using [senderSs58Account](https://github.com/AstarNetwork/astar-apps/blob/cbf4e5688c910f366337d5d0cc1a072013120820/src/hooks/useAccount.ts#L24) instead of `currentAccount`, so that mapped SS58 account is used in every logic.  
  * e2e tests are not applied until resolving chopstick issues (connecting EVM network) 

**Check list**

- [ ] contains breaking changes
- [x] adds new feature
- [ ] modifies existing feature (bug fix or improvements)
- [ ] relies on other tasks
- [ ] documentation changes
- [ ] tested on mobile devices

**Adds**

image

### CODE REVIEW

File: src/v2/services/implementations/DappStakingService.ts [hunk 1]
```diff
@@ -91,13 +104,69 @@ export class DappStakingService implements IDappStakingService {
...
     Guard.ThrowIfUndefined('stakerAddress', stakerAddress);
+    const isH160 = isValidEvmAddress(stakerAddress);
 
-    const stakeCall = await this.dappStakingRepository.getBondAndStakeCall(contractAddress, amount);
-    await this.wallet.signAndSend({
-      extrinsic: stakeCall,
-      senderAddress: stakerAddress,
-      successMessage,
-    });
+    if (isH160) {
```
Comment (bobo-k2) [inline src/v2/services/implementations/DappStakingService.ts:109]: I suggest to create another dapp staking repository
`export class EvmDappStakingService implements IDappStakingService`
 which will handle EVM calls and inject it to the portal if user chooses Metamask (a wallet services are handled the same way). By doing this you will keep `DappStakingService` clean and simple.

File: src/v2/services/implementations/DappStakingService.ts [hunk 2]
```diff
@@ -91,13 +104,69 @@ export class DappStakingService implements IDappStakingService {
...
+      const rawTx = {
+        nonce,
+        gasPrice: web3.utils.toHex(gasPrice),
+        from: stakerAddress,
+        to: dappStakingContract,
+        value: '0x0',
+        data: contract.methods.bond_and_stake(contractAddress, amount).encodeABI(),
+      };
+      const estimatedGas = await web3.eth.estimateGas(rawTx);
+      await web3.eth
```
Comment (bobo-k2) [inline src/v2/services/implementations/DappStakingService.ts:133]: Signing should be handled by `MetamaskWalletService`

**Comment by impelcrypto** [comment]
@bobo-k2 Thank you for checking, just let you know that I still need to work a lot in this branch :D

File: src/hooks/dapps-staking/useCompoundRewards.ts [hunk 3]
```diff
@@ -76,9 +78,11 @@ export function useCompoundRewards() {
 
   const setRewardDestination = async (rewardDestination: RewardDestination): Promise<void> => {
     try {
-      const dappStakingService = container.get<IDappStakingService>(Symbols.DappStakingService);
+      const dappStakingService = container.get<IDappStakingService>(
```
Comment (bobo-k2) [inline src/hooks/dapps-staking/useCompoundRewards.ts:81]: Instead having this code on multiple you can do something similar we did for `IWalletService`
```
container.bind<interfaces.Factory<IWalletService>>(Symbols.WalletFactory).toFactory(() => {
    return () => {
      return container.get<IWalletService>(currentWalletType);
    };
  });
```
After that you can use factory to get proper instance of service

**Review by bobo-k2 — CHANGES_REQUESTED** [review]
Great work Roy, I have one general comment, e2e tests are missing :)

**Comment by impelcrypto** [comment]
@bobo-k2 

>  I have one general comment, e2e tests are missing :)

Yes, I've mentioned this on `Pull Request Summary`. I will add it once the issue has been resolved.
> * e2e tests are not applied until resolving chopstick issues (connecting EVM network)

File: src/hooks/dapps-staking/useCompoundRewards.ts (hunk 3 above)
Comment (impelcrypto) [reply src/hooks/dapps-staking/useCompoundRewards.ts]: Thanks, I fixed it in [this](https://github.com/AstarNetwork/astar-apps/pull/857/commits/53e2951cf9c6eb8f63b673304086e82c493479ce) commit.

**Review by bobo-k2 — APPROVED** [review]
Great work LGTM

**Review by gluneau — APPROVED** [review]
It works!

**Comment by impelcrypto** [comment]
I'll merge this PR once I get confirmation from @Kahonnohak

**Review by Kahonnohak — APPROVED** [review]
LGTM!

**Pull request merged**